File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.cpp

#include <bits/stdc++.h>
using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> answer;

    vector<pair<int,int>> v;
    int sum = 0;
    int i = 0; int j =0;

    sum += sequence[0];


    while(i < sequence.size()){
        if(sum > k){
            sum -= sequence[j];
            j++;
        }
        else if( k > sum ){
            i++;
            sum += sequence[i];
        }
        else if ( sum == k ){
            v.push_back({j,i});

            i++;
            sum += sequence[i];
        }
    }
    int min1 = 0; int min2 = INT_MAX;
    for(auto t : v){
        if( min2 - min1 > t.second - t.first){
            min2 = t.second; min1 = t.first;
        }
    }
    //cout << min1 << " " << min2;

    answer.push_back(min1); answer.push_back(min2);
    return answer;
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.js
function solution(sequence, k) {
  var answer = [];
 
  let left = 0;
  let right = 0;
  let sum = sequence[left];

  while(right < sequence.length){

    if(sum === k){
      let ansLength = answer[1] - answer[0];

      if(!answer.length){
        answer = [left, right];
      }
      else if(ansLength > right - left){
        answer = [left, right];
      }
    }

    if(sum <= k){
      right++;
      sum+=sequence[right];
    }else if(left<=right){
      sum-=sequence[left];
      left++;
      
    }

  }
 

  return answer;
}